### ISSUE TITLE
AES tests misconfigure one-hot fields of a transaction

### ISSUE THREAD

**Issue by drewmacrae** [lowRISC/opentitan#15642]
Many aes tests including the aes_smoketest and the entropy_src_edn_reqs_test configure the mask reseed field to a reserved value.

Because the transaction type is a c struct, the field is initialized to 0 and then unset and is therefore not a 1-hot value as it's supposed to be. Adding a check to the dif makes these tests fail. Initializing the value will fix them (and should also make them run faster)

> estimate 4

**Comment by vogelpi** [comment]
Hi @drewmacrae ,
the reason for using 1 hot encodings here is security. Internally, the hardware will resolve invalid values to the safest setting (i.e., reseed on every block).

I just filed a PR that touches all these tests because it changes some bits in the control register. Let me quickly extend the PR to set a valid configuration for this field in every test.

**Comment by vogelpi** [comment]
Here is the PR #15641, I've now modified these tests such they all set a valid value.

**Comment by drewmacrae** [comment]
Should we have a sort of "one-hot" macro that lets us enumerate and check one-hot types like this to avoid similar escapes? In general non-contiguous enums are problematic. For these things with reserved values are we interested in helping test/sw authors to configure things properly?

**Comment by drewmacrae** [comment]
We could emit something like the following function
```
/**
 * Test if AES Mode is Valid
 *
 * Check that aes_mode_t is enumerated in dif_aes_mode
 * @param A dif_aes_mode_t
 * @return The variable has a valid value
 */
bool dif_aes_valid_mode(dif_aes_mode_t value);
```

**Comment by johngt** [comment]
@vogelpi - just tagging you here to see if this is still a valid issue for AES/M2?
If so - leave the issue open and we can keep it in the AES project + we will need to estimate effort. Otherwise we should add it to the backlog and de-prioritise accordingly.

**Comment by drewmacrae** [comment]
@alphan pointed out there are builtins in the bitfield library. So I'll add checks to the dif and file a PR with fixes for anything that's still broken after #15641 gets merged.

**Comment by vogelpi** [comment]
Hi @drewmacrae ,
this sounds good to me. FYI #15641 has now been merged.

@johngt , this is not actually related to the hardware or AES/M2 effort. I'll try to move this out of the corresponding view.

**Comment by drewmacrae** [comment]
I'd like t0 make the aes_smoketest reseed less often to get better coverage and performance.

**Comment by msfschaffner** [comment]
Ok sounds like this is not really needed for M2, so I'll relabel it to M3. Please correct if that is wrong.

**Comment by vogelpi** [comment]
> I'd like to make the aes_smoketest reseed less often to get better coverage and performance.

Hey @drewmacrae , the `aes_smoketest` just encrypts and decrypts 1 block each. Currently, it should be doing 5 reseeding operations:
1. Right after coming out of reset. This cannot be changed.
2. After writing the encryption key. This can be changed by setting `AUX_CONTROL_SHADOWED.KEY_TOUCH_FORCES_RESEED` to 0.
3. Before finishing the encryption of the block. This can be changed by setting `CTRL_SHADOWED.PRNG_RESEED_RATE` to 0x2.
4. After writing the decryption key. Again can be changed by `AUX_CONTROL_SHADOWED.KEY_TOUCH_FORCES_RESEED` to 0.
5. Before finishing the decryption of the block. Again this can be changed by setting `CTRL_SHADOWED.PRNG_RESEED_RATE` to 0x2.

I would suggest to either remove the reseed after 2 and 4 or after 3 and 5.

**Comment by vogelpi** [comment]
This issue discusses options to speed up tests (check for 1-hot encodings inside the DIF, reducing reseed operations in AES TLTs). This would be nice to have for M2.5 but I believe we can do without it - P3.

**Comment by moidx** [comment]
Closing as obsolete.  If there is still an issue lets re-open.
